In [1]:
import numpy as np
import pandas as pd

In [2]:
data = pd.read_csv("Data_for_UCI_named.csv")

In [3]:
data.head()

tau1      tau2      tau3      tau4        p1        p2        p3  \
0  2.959060  3.079885  8.381025  9.780754  3.763085 -0.782604 -1.257395   
1  9.304097  4.902524  3.047541  1.369357  5.067812 -1.940058 -1.872742   
2  8.971707  8.848428  3.046479  1.214518  3.405158 -1.207456 -1.277210   
3  0.716415  7.669600  4.486641  2.340563  3.963791 -1.027473 -1.938944   
4  3.134112  7.608772  4.943759  9.857573  3.525811 -1.125531 -1.845975   

         p4        g1        g2        g3        g4      stab     stabf  
0 -1.723086  0.650456  0.859578  0.887445  0.958034  0.055347  unstable  
1 -1.255012  0.413441  0.862414  0.562139  0.781760 -0.005957    stable  
2 -0.920492  0.163041  0.766689  0.839444  0.109853  0.003471  unstable  
3 -0.997374  0.446209  0.976744  0.929381  0.362718  0.028871  unstable  
4 -0.554305  0.797110  0.455450  0.656947  0.820923  0.049860  unstable

In [4]:
data = data.drop(["stab"], 1)

In [5]:
from sklearn.model_selection import train_test_split
X = data.drop(["stabf"], 1)
y = data["stabf"]

In [6]:
 X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=1)

In [7]:
# using standard scaler to scale the features

from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
normalised_train_df = scaler.fit_transform(X_train)
normalised_train_df = pd.DataFrame(normalised_train_df, columns=X_train.columns)

normalised_test_df = scaler.transform(X_test)
normalised_test_df = pd.DataFrame(normalised_test_df, columns=X_test.columns)

### Question 14
### What is the accuracy on the test set using the random forest classifier? In 4 decimal places.

In [8]:
#importing RandomForestClassifier

from sklearn.ensemble import RandomForestClassifier
rfc = RandomForestClassifier(random_state = 1)
rfc.fit(normalised_train_df, y_train)
rfc_pred = rfc.predict(normalised_test_df)

In [9]:
# classification report is used to get all the metrics in a table

from sklearn.metrics import classification_report
print(classification_report(y_test, rfc_pred, digits = 4))

              precision    recall  f1-score   support

      stable     0.9191    0.8778    0.8980       712
    unstable     0.9341    0.9573    0.9456      1288

    accuracy                         0.9290      2000
   macro avg     0.9266    0.9176    0.9218      2000
weighted avg     0.9288    0.9290    0.9286      2000



In [10]:
# or we could import accuracy score directly
from sklearn.metrics import accuracy_score
accuracy = accuracy_score(y_true = y_test, y_pred = rfc_pred)
print(round(accuracy, 4))

0.929


### Question 15
### What is the accuracy on the test set using the xgboost classifier? In 4 decimal places.

In [11]:
# importing xgboost
from xgboost import XGBClassifier
xgb = XGBClassifier(random_state=1)
xgb.fit(normalised_train_df, y_train)
xgb_pred = xgb.predict(normalised_test_df)


# for the classificaton report
print(classification_report(y_test, xgb_pred, digits = 4))

              precision    recall  f1-score   support

      stable     0.9351    0.9101    0.9224       712
    unstable     0.9510    0.9651    0.9580      1288

    accuracy                         0.9455      2000
   macro avg     0.9430    0.9376    0.9402      2000
weighted avg     0.9453    0.9455    0.9453      2000



In [12]:
accuracy = accuracy_score(y_true = y_test, y_pred = xgb_pred)
print(round(accuracy, 4))

0.9455


### Question 16
### What is the accuracy on the test set using the LGBM classifier? In 4 decimal places.

In [13]:
# importing lightgbm classifier
from lightgbm import LGBMClassifier
lgbm = LGBMClassifier(random_state=1)
lgbm.fit(normalised_train_df, y_train)
lgbm_pred = lgbm.predict(normalised_test_df)

In [14]:
print(classification_report(y_test, lgbm_pred, digits = 4))

              precision    recall  f1-score   support

      stable     0.9297    0.8919    0.9104       712
    unstable     0.9415    0.9627    0.9520      1288

    accuracy                         0.9375      2000
   macro avg     0.9356    0.9273    0.9312      2000
weighted avg     0.9373    0.9375    0.9372      2000



In [15]:
accuracy = accuracy_score(y_true = y_test, y_pred = lgbm_pred)
print(round(accuracy, 4))

0.9375


### Question 17
### To improve the Extra Trees Classifier, you will use the following parameters (number of estimators, minimum number of samples, minimum number of samples for leaf node and the number of features to consider when looking for the best split) for the hyperparameter grid needed to run a Randomized Cross Validation Search (RandomizedSearchCV). 

In [16]:
#importing extra tree classifier and then initializing it

from sklearn.ensemble import ExtraTreesClassifier
tree = ExtraTreesClassifier(random_state =1)
tree.fit(normalised_train_df, y_train)
tree_pred = tree.predict(normalised_test_df)

In [17]:
print(classification_report(y_test, tree_pred, digits = 4))

              precision    recall  f1-score   support

      stable     0.9410    0.8511    0.8938       712
    unstable     0.9218    0.9705    0.9455      1288

    accuracy                         0.9280      2000
   macro avg     0.9314    0.9108    0.9197      2000
weighted avg     0.9287    0.9280    0.9271      2000



In [18]:
accuracy = accuracy_score(y_true = y_test, y_pred = tree_pred)
print(round(accuracy, 4))

0.928


In [19]:
n_estimators = [50, 100, 300, 500, 1000]
min_samples_split = [2, 3, 5, 7, 9]
min_samples_leaf = [1, 2, 4, 6, 8]
max_features = ['auto', 'sqrt', 'log2', None] 
hyperparameter_grid = {'n_estimators': n_estimators,
                       'min_samples_leaf': min_samples_leaf,
                       'min_samples_split': min_samples_split,
                       'max_features': max_features}

In [20]:
from sklearn.model_selection import RandomizedSearchCV
clf = RandomizedSearchCV(tree, hyperparameter_grid, random_state=1)
check = clf.fit(normalised_train_df, y_train)

### What are the best hyperparameters from the randomized search CV?

In [21]:
check.best_params_.values()

dict_values([1000, 2, 8, None])

### Question 18
### Train a new ExtraTreesClassifier Model with the new Hyperparameters from the RandomizedSearchCV (with random_state = 1). Is the accuracy of the new optimal model higher or lower than the initial ExtraTreesClassifier model with no hyperparameter tuning?


In [22]:
#Training a new ExtraTreesClassifier Model with the new Hyperparameters from the RandomizedSearchCV

tree_hype = ExtraTreesClassifier(n_jobs = -1, verbose = 1,
                                  n_estimators=1000, min_samples_split=2, 
                                 min_samples_leaf=8, max_features=None, random_state=1)
tree_hype.fit(normalised_train_df, y_train)
tree_hype_pred = tree_hype.predict(normalised_test_df)

[Parallel(n_jobs=-1)]: Using backend ThreadingBackend with 4 concurrent workers.
[Parallel(n_jobs=-1)]: Done  42 tasks      | elapsed:    0.2s
[Parallel(n_jobs=-1)]: Done 192 tasks      | elapsed:    1.2s
[Parallel(n_jobs=-1)]: Done 442 tasks      | elapsed:    2.8s
[Parallel(n_jobs=-1)]: Done 792 tasks      | elapsed:    5.1s
[Parallel(n_jobs=-1)]: Done 1000 out of 1000 | elapsed:    6.5s finished
[Parallel(n_jobs=4)]: Using backend ThreadingBackend with 4 concurrent workers.
[Parallel(n_jobs=4)]: Done  42 tasks      | elapsed:    0.0s
[Parallel(n_jobs=4)]: Done 192 tasks      | elapsed:    0.1s
[Parallel(n_jobs=4)]: Done 442 tasks      | elapsed:    0.3s
[Parallel(n_jobs=4)]: Done 792 tasks      | elapsed:    0.5s
[Parallel(n_jobs=4)]: Done 1000 out of 1000 | elapsed:    0.6s finished


In [23]:
print(classification_report(y_test, tree_hype_pred, digits = 4))

              precision    recall  f1-score   support

      stable     0.9211    0.8694    0.8945       712
    unstable     0.9300    0.9589    0.9442      1288

    accuracy                         0.9270      2000
   macro avg     0.9256    0.9141    0.9193      2000
weighted avg     0.9268    0.9270    0.9265      2000



In [24]:
accuracy = accuracy_score(y_true = y_test, y_pred = tree_hype_pred)
print(round(accuracy, 4))

0.927


###  The accuracy of the new optimal model is lower(0.927) than the initial ExtraTreesClassifier model with no hyperparameter tuning(0.928)

### Question 20
### Find the feature importance using the optimal ExtraTreesClassifier model. Which features are the most and least important respectively?

In [25]:
tree_hype.feature_importances_

array([0.13723975, 0.1405075 , 0.13468029, 0.13541676, 0.00368342,
       0.00533686, 0.00542927, 0.00496249, 0.10256244, 0.10757765,
       0.11306268, 0.10954089])

In [26]:
importance = tree.feature_importances_
df = pd.DataFrame(importance, index = normalised_test_df.columns)

In [27]:
df = df.sort_values(by = 0)

In [28]:
df

0
p1    0.039507
p2    0.040371
p4    0.040579
p3    0.040706
g1    0.089783
g2    0.093676
g4    0.094019
g3    0.096883
tau3  0.113169
tau4  0.115466
tau1  0.117397
tau2  0.118445

### The least important feature is p1
### The most important feature is tau2